src/cmd/run.js

```diff
@@ -63,23 +68,48 @@ export function defaultWatcherCreator(
     sourceDir,
     artifactsDir,
     onChange: () => {
-      log.debug(`Reloading add-on ID ${addonId}`);
-
-      return client.reloadAddon(addonId)
-        .catch((error) => {
-          log.error('\n');
-          log.error(error.stack);
-          desktopNotifications({
-            title: 'web-ext run: error occured',
-            message: error.message,
-          });
-          throw error;
-        });
+      return addonReload({addonId, client});
     },
     shouldWatchFile: (file) => fileFilter.wantFile(file),
   });
 }
 
+export type ReloadParams = {
+  addonId: string,
+  client: RemoteFirefox,
+  desktopNotifications?: typeof defaultDesktopNotifications,
+};
+
+export function defaultAddonReload(
+  {
+  addonId, client, desktopNotifications = defaultDesktopNotifications,
+ }: ReloadParams
+): Promise<void> {
+  log.debug(`Reloading add-on ID ${addonId}`);
+  return client.reloadAddon(addonId)
+    .catch((error) => {
+      log.error('\n');
+      log.error(error.stack);
+      desktopNotifications({
+        title: 'web-ext run: error occured',
+        message: error.message,
+      });
+      throw error;
+    });
+}
+
+export type ExitParams = {
+  logger?: typeof log,
+  killer?: typeof process.kill,
+};
+
+export async function defaultExitProgram(
+  {logger = log, killer = process.kill}: ExitParams = {},
+): Promise<void> {
+  logger.info('Exiting web-ext on user request');
+  killer(process.pid, 'SIGINT');
```

Comment:
@saintsebastian I did some testing on Windows (and some additional reading about the "how to cleanly exit a nodejs app") and it seems that the current implementation (the one based on `SIGINT`) has a couple of issues:

- web-ext does not exit anymore when the Firefox instance started by web-ext is closed (this is probably related to the new listener attached to the `process.stdin`, and it happens on every platform, I directly tried only on Linux and Windows, but given the kind of issue OSX has probably the same issue)

- on Windows, web-ext does not remove the "temporarely created profile" anymore (it does on Linux, and probably also on OSX)

The above "Windows specific issue" seems to be related to the `"SIGINT"` strategy, which probably doesn't behave correctly on Window (e.g. it exits the process buy the `"SIGINT"` event is probably missing, and the `"firefox-profile"` npm dependency subscribe a listener to the `"exit"` and `"SIGINT"` events on the `process` object to be able to cleanup the "temporarily generated profile")

I did some further experimentations on Windows and it seems that if web-ext exits using the `process.exit` method, the `"exit"` event is correctly received on Windows (as well as on Linux and OSX).

Unfortunately, on Windows if the Firefox instance is still running, `"firefox-profile"` fails to remove the generate temporary profile dir (because of an `"EBUSY: resource busy or locked, unlink 'C:\Users\..."` error).

The following is the only workaround that I found to fix both the above issues:

- define a global module var (e.g. `let userExitRequest = false;`) to be able to recognize when a Firefox instance is exiting because of an explicit user request

- at line 312, when Ctrl-C shortcut has been received, set to true the defined global module var (e.g. `userExitRequest = true;`) to recognize that we are killing the running Firefox instance because of a user request, and then kill the Firefox instance using `runningFirefox.kill()`

- at line 143, when the "close" event has been emitted by `firefoxProcess`, check if the user has requested to exit and in that case use `process.exit(0)` to exit it (with a "success" process exit value), on the contrary if the Firefox instance has been closed for unknown reasons we should probably use `process.exit(1)` (Unfortunately it seems that we need to force the nodejs process in both the scenarios, or web-ext will never exit if it is running in "interactive mode" and the Firefox instance closes itself without any user interventions).

By killing the running Firefox instance, and exiting the nodejs process once it has been closed, it seems that Windows is able to unlink the temporary generated profile as expected.

Note: the main problem with using `process.exit` to exit a nodejs application seems to be related to the flushing of the stdout/stderr data, that is asynchronous.

@kumar303 how that sounds to you?